### ISSUE TITLE
Type error when stateful

### ISSUE THREAD

**Issue by ekerazha** [fchollet/keras#1130]
After the fix for issue  #1125, I'm trying

    in_out_neurons = 1
    hidden_neurons = 10
    batch_size = 10
    
    model = Sequential()
    model.add(GRU(hidden_neurons, batch_input_shape=(batch_size, 5, in_out_neurons), return_sequences=True, stateful=True))

but I get this error

    TypeError: data type not understood

This is the traceback:

    model.add(GRU(hidden_neurons, batch_input_shape=(batch_size, 5, in_out_neurons), return_sequences=True, stateful=True))

    recurrent.py", line 190, in __init__
    super(GRU, self).__init__(**kwargs)

    recurrent.py", line 25, in __init__
    super(Recurrent, self).__init__(**kwargs)

    core.py", line 30, in __init__
    self.set_input_shape(tuple(kwargs['batch_input_shape']))

    core.py", line 95, in set_input_shape
    self.build()

    recurrent.py", line 219, in build
    self.states = [K.zeros(input_shape[0], self.output_dim)]

    theano_backend.py", line 80, in zeros
    return variable(np.zeros(shape), dtype, name)

    theano_backend.py", line 33, in variable
    value = np.asarray(value, dtype=dtype)

    numeric.py", line 474, in asarray
    return array(a, dtype, copy=False, order=order)

Theano is up-to-date. Am I missing anything?

**Comment by futurely** [comment]
Does the tensorflow backend work?

**Comment by ekerazha** [comment]
I can't test TensorFlow at the moment.

**Comment by fchollet** [comment]
Should be fixed now. Will require more unit tests. Let me know how it goes.

**Comment by RaffEdwardBAH** [comment]
@fchollet I just updated (as of a few minutes ago) and am still getting errors 

```python
model = Sequential()
model.add(Embedding(tokens, 128))
model.add(LSTM(1024, return_sequences=True, stateful=True, batch_input_shape=(batch_size,max_len-1, 128) ))
model.add(Dropout(0.5))
model.add(LSTM(1024, return_sequences=True, stateful=True, batch_input_shape=(batch_size,max_len-1, 1024) ))
model.add(Dropout(0.5))
model.add(LSTM(1024, return_sequences=True, stateful=True, batch_input_shape=(batch_size,max_len-1, 1024) ))
model.add(Dropout(0.5))
model.add(TimeDistributedDense(tokens))
model.add(Activation('softmax'))

model.compile(loss='categorical_crossentropy', optimizer=Adam(clipnorm=20))
```

Gives the error 
```
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-8-893d09bc3dcd> in <module>()
      1 model = Sequential()
      2 model.add(Embedding(tokens, 128))
----> 3 model.add(LSTM(1024, return_sequences=True, stateful=True, batch_input_shape=(batch_size,max_len-1, 128) ))
      4 model.add(Dropout(0.5))
      5 model.add(LSTM(1024, return_sequences=True, stateful=True, batch_input_shape=(batch_size,max_len-1, 1024) ))

/usr/local/lib/python2.7/dist-packages/Keras-0.3.0-py2.7.egg/keras/layers/containers.pyc in add(self, layer)
     30         self.layers.append(layer)
     31         if len(self.layers) > 1:
---> 32             self.layers[-1].set_previous(self.layers[-2])
     33             if not hasattr(self.layers[0], 'input'):
     34                 self.set_input()

/usr/local/lib/python2.7/dist-packages/Keras-0.3.0-py2.7.egg/keras/layers/core.pyc in set_previous(self, layer, connection_map)
     42             assert self.supports_masked_input(), "Cannot connect non-masking layer to layer with masked output"
     43         self.previous = layer
---> 44         self.build()
     45 
     46     def build(self):

/usr/local/lib/python2.7/dist-packages/Keras-0.3.0-py2.7.egg/keras/layers/recurrent.pyc in build(self)
    295                 raise Exception('If a RNN is stateful, a complete ' +
    296                                 'input_shape must be provided ' +
--> 297                                 '(including batch size).')
    298             self.states = [K.zeros((input_shape[0], self.output_dim)),
    299                            K.zeros((input_shape[0], self.output_dim))]

Exception: If a RNN is stateful, a complete input_shape must be provided (including batch size).
```
This is the same problem I've blocked on in issue #98

**Comment by ekerazha** [comment]
@fchollet 
The issue that I reported is fixed for me, so I'll close this issue. However there's another issue, so I'm going to open a new bug report.
